mmdet3d/models/dense_heads/shape_aware_head.py

```diff
@@ -0,0 +1,640 @@
+
+
+@HEADS.register_module()
+class ShapeAwareHead(Anchor3DHead):
+    """Shape-aware grouping head for SSN.
+
+    Args:
+        tasks (dict): Shape-aware groups of multi-class objects.
+        assign_per_class (bool): Whether to do assignment for each class.
+            Default: True.
+        kwargs (dict): Other arguments are the same as those in \
+            :class:`Anchor3DHead`.
+    """
+
+    def __init__(self, tasks, assign_per_class=True, **kwargs):
+        self.tasks = tasks
+        self.featmap_sizes = []
+        super().__init__(assign_per_class=assign_per_class, **kwargs)
+
+    def _init_layers(self):
+        """Initialize neural network layers of the head."""
+        self.heads = nn.ModuleList()
+        cls_ptr = 0
+        for task in self.tasks:
+            sizes = self.anchor_generator.sizes[cls_ptr:cls_ptr +
+                                                task['num_class']]
+            num_size = torch.tensor(sizes).reshape(-1, 3).size(0)
+            num_rot = len(self.anchor_generator.rotations)
+            num_base_anchors = num_rot * num_size
+            branch = dict(
+                type=task['type'],
+                num_cls=self.num_classes,
+                num_base_anchors=num_base_anchors,
+                box_code_size=self.box_code_size,
+                in_channels=self.in_channels)
+            self.heads.append(build_head(branch))
+            cls_ptr += task['num_class']
+
+    def init_weights(self):
+        """Initialize the weights of head."""
+        for head in self.heads:
+            head.init_weights()
+
+    def forward_single(self, x):
+        """Forward function on a single-scale feature map.
+
+        Args:
+            x (torch.Tensor): Input features.
+        Returns:
+            tuple[torch.Tensor]: Contain score of each class, bbox \
+                regression and direction classification predictions.
+        """
+        results = []
+
+        for head in self.heads:
+            results.append(head(x))
+
+        cls_score = torch.cat([result['cls_score'] for result in results],
+                              dim=1)
+        bbox_pred = torch.cat([result['bbox_pred'] for result in results],
+                              dim=1)
+        dir_cls_preds = None
+        if self.use_direction_classifier:
+            dir_cls_preds = torch.cat(
+                [result['dir_cls_preds'] for result in results], dim=1)
+
+        self.featmap_sizes = []
+        for i, task in enumerate(self.tasks):
+            for _ in range(task['num_class']):
+                self.featmap_sizes.append(results[i]['featmap_size'])
+        assert len(self.featmap_sizes) == len(self.anchor_generator.ranges), \
+            'Length of feature map sizes must be equal to length of ' + \
+            'different ranges of anchor generator.'
+
+        return cls_score, bbox_pred, dir_cls_preds
+
+    def get_anchors(self, input_metas, device='cuda'):
```

Comment:
Do we need to rewrite the function `get_anchors'?

Reply:
Good idea. The previous get_anchors in the reformatting procedure can not be unified with the original one, but the current version allows. Already updated.